### ISSUE TITLE
Iterator should not modify the heap (it should use nlargest(length) instead of pop)

### ISSUE THREAD

**Issue by supriyo-biswas** [ignlg/heap-js#643]
Currently, this library implements the heap with the `pop()` method, which causes the heap to be emptied on a single iteration. For example:

```
import { Heap } from 'heap-js'

const heap = new Heap()
heap.push(10)
heap.push(5)
heap.push(15)
heap.push(3)

// prints 3, 5, 10, 15
for (const item of heap) {
  console.log(item)
}

// does not print anything
for (const item of heap) {
  console.log(item)
}
```

As it is not a standard behaviour of iterators to modify the underlying data structure, the `sorted(nlargest(length))` method should be used instead of `pop()`.

**Comment by ignlg** [comment]
Good point @supriyo-biswas , thanks for opening this issue. I will fix it asap.

**Comment by ignlg** [comment]
Indeed, for the iterator to return heap-sorted results, the only possible approach would be to clone the heap internally and consume the clone.

Furthermore, it is not trivial to walk a binary tree linearly, isn't it? By definition. The only way to iterate over a heap that works as a heap (popping the root value and replacing it with a leaf) is to consume it.

I could be biased by Rust dev, so I will take a look at other languages implementations to find the best approach, because `sorted(nlargest(length))` will convert the entire dataset into an array and then sorting it, invalidating any benefit of using a heap.

**Comment by ignlg** [comment]
Fixed on version `v2.2.0`, already published.

**Comment by ignlg** [comment]
Friendly reminder that you can always [clone the heap](https://ignlg.github.io/heap-js/classes/Heap.html#clone) and consume the clone:

```typescript
const heap = new Heap();
heap.init([1,23,45,6]);

const cloned = heap.clone();
for (const topValue of cloned) {
  // Do something
}

assert(cloned.length === 0);
assert(heap.length === 4);
```
